waiter/src/waiter/state.clj

```diff
@@ -31,44 +31,86 @@
+  [request-id->use-reason-map earliest-request-threshold-time expired-instances? starting-instances?
+   {:keys [slots-used status-tags] :as state}]
+  (and (or (and (zero? slots-used)
+                (not-any? #(contains? status-tags %) [:killed :locked]))
+           (and (expired? state)
+                (only-lingering-requests? request-id->use-reason-map earliest-request-threshold-time)))
+       (or (not (and expired-instances? (contains? status-tags :healthy)))
+           (some #(contains? status-tags %) [:blacklisted :expired]))
+       (or (not (and expired-instances? starting-instances?))
+           (and (contains? status-tags :unhealthy)
+                (not (contains? status-tags :starting))))))
+
+(defn- find-max
+  "Uses the provided comparison function to find the max element in collection."
+  [comparison-fn collection]
+  (reduce #(if (pos? (comparison-fn %1 %2)) %1 %2) (first collection) (rest collection)))
+
+(defn- expired-instances?
+  "Returns truthy value if there is some instance in an expired state."
+  [instance-id->state]
+  (some (fn [[_ {:keys [status-tags]}]] (contains? status-tags :expired)) instance-id->state))
+
+(defn- starting-instances?
+  "Returns truthy value if there is some instance in an starting state."
+  [instance-id->state]
+  (some (fn [[_ {:keys [status-tags]}]] (contains? status-tags :starting)) instance-id->state))
 
 (defn find-killable-instance
   "While killing instances choose using the following logic:
    if there are expired and starting instances, only kill unhealthy instances that are not starting;
    choose instances in following order:
    - choose amongst idle instances
-   - choose youngest instance to break ties
+   - choose youngest healthy instance or oldest expired instance to break ties
    - choose idle unhealthy ones with highest priority
    - next, choose healthy, but blacklisted instances
-   - next, choose expired instance
+   - next, choose expired instances (idle or processing only lingering requests)
    - finally, choose healthy instances."
-  [id->instance instance-id->state acceptable-instance-id?]
-  (let [find-max (fn [comparison-fn collection]
-                   (reduce
-                     #(if (pos? (comparison-fn %1 %2)) %1 %2)
-                     (first collection)
-                     (rest collection)))
-        instance-id-state-pair->categorizer-vec (fn [[_ {:keys [status-tags]}]]
+  [id->instance instance-id->state acceptable-instance-id? instance-id->request-id->use-reason-map lingering-request-threshold-ms]
+  (let [earliest-request-threshold-time (t/minus (t/now) (t/millis lingering-request-threshold-ms))
+        instance-id-state-pair->categorizer-vec (fn [[instance-id {:keys [slots-used status-tags] :as state}]]
                                                   ; most important goes first
                                                   (vector
```

Comment:
Nitpick: Use vector literal

Reply:
I prefer `vector` as the elements are spread over multiple lines.